test/reconstruct-test.js

```diff
@@ -0,0 +1,143 @@
+const vows = require('vows');
+const assert = require('assert');
+const ret = require('../dist');
+const { tokenToString } = require('typescript');
+const reconstruct = require('../dist/reconstruct').reconstruct
+const partialConstruct = require('../dist/reconstruct').partialConstruct
+const tokens = require('../dist/types').types
+
+const inverseTestFactory = (regexp) => {
+  return {
+    topic: ret(regexp),
+
+    [`Checking ${regexp} reconstructs`]: (t) => {
+      const reconstructed = reconstruct(t) // May need to do some sort of sanitisation here
+      assert.isString(reconstructed);
+      assert.deepStrictEqual(reconstructed, regexp.replace('[^0-9]', '\\D'))
+    },
+
+    [`Checking ${regexp} reconstructs using partialConstruct`]: (t) => {
+      const reconstructed = partialConstruct(t) // May need to do some sort of sanitisation here
+      assert.isString(reconstructed);
+      assert.deepStrictEqual(reconstructed, regexp.replace('[^0-9]', '\\D'))
```

Comment:
this is only needed for one test below right? is similar replacement needed for all of the predefined sets?

this also brings up another point, the reconstruction isn't 1 to 1 with custom sets like this that match a predefined set, or custom repetitions that match predefined reptition tokens (i.e. `a{0,1}` will become `a+`). is this something that can be avoided with the details PR?

Reply:
In the reconstruction all predefined sets are simplified, see https://github.com/jeswr/ret.js/blob/3973e463f5c6fa03ec4fccc22b0a1ad1daa4d667/lib/reconstruct.ts#L4-L12 and https://github.com/jeswr/ret.js/blob/3973e463f5c6fa03ec4fccc22b0a1ad1daa4d667/lib/reconstruct.ts#L44.

Most of the test cases just steal the regular expressions in the `main-test` file and make sure that the inversion property holds. So it is just that there are no tests that contained the 'unsimplified' versions of any of the other sets. I will add some test cases for this.

I don't think that there is then really an issue in the details PR as all of the details (other than the reconstructed regular expression) are concerned about properties of a string that is being validated against the regular expression rather than properties about the appearance of the regular expressions itself.

Reply:
Similarly all custom repetitions that can be simplified are simplfied https://github.com/fent/ret.js/blob/755e315078423097a0c59914db651265a44ade6b/lib/reconstruct.ts#L61-L70